compiler/noirc_evaluator/src/ssa/opt/defunctionalize.rs

```diff
@@ -226,19 +226,22 @@ fn find_variants(ssa: &Ssa) -> Variants {
         );
     }
 
-    let mut signature_to_functions_as_value: BTreeMap<Signature, Vec<FunctionId>> = BTreeMap::new();
+    let mut signature_to_functions_as_value: Variants = BTreeMap::new();
 
     for function_id in functions_as_values {
-        let signature = ssa.functions[&function_id].signature();
-        signature_to_functions_as_value.entry(signature).or_default().push(function_id);
+        let func = &ssa.functions[&function_id];
+        let signature = func.signature();
+        let runtime = func.runtime();
+        signature_to_functions_as_value.entry((signature, runtime)).or_default().push(function_id);
     }
 
     let mut variants: Variants = BTreeMap::new();
 
     for (dispatch_signature, caller_runtime) in dynamic_dispatches {
-        let target_fns =
-            signature_to_functions_as_value.get(&dispatch_signature).cloned().unwrap_or_default();
-        variants.insert((dispatch_signature, caller_runtime), target_fns);
+        let key = (dispatch_signature, caller_runtime);
+        let target_fns = signature_to_functions_as_value.get(&key).cloned().unwrap_or_default();
```

Comment:
With this change we end up in a situation where if the lambda is created in constrained code, and subsequently is passed to unconstrained code, then this list is empty and the compiler crashes further down.